### ISSUE TITLE
Document the meaning of trial's terminal output stanza

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7175]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)| @warner reported|
|-|-|
|Trac ID|trac#1384|
|Type|task|
|Created|2005-12-30 13:39:00Z|
|Branch|https://github.com/twisted/twisted/tree/doc-trial-output-1384|

The trial man page and the various howtos which discuss trial should explain the meaning of the output, so that people do not have to guess.

Attachments:

* [test_warner.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/589/58907516874aa70188525b2518faf0e285fd821a/65ceb793be842d997bda141487d87f3d9467409a.py) (302 bytes) - added by warner on 2005-12-30 13:39:00Z - 
* [trial-output.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/589/58907516874aa70188525b2518faf0e285fd821a/56e58dfe19c2a3dadccf7215857865e7640115ee.txt) (3282 bytes) - added by jml on 2006-01-03 11:44:20Z - 
* [document-trial-output-1384.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/589/58907516874aa70188525b2518faf0e285fd821a/b66188adee164eef87a553572848d0f3a92591d1.patch) (3085 bytes) - added by Screwtape on 2010-06-29 14:38:21Z - Document trial's output stanza in the manpage.

<details><summary>Searchable metadata</summary>

```
trac-id__1384 1384
type__task task
reporter__warner warner
priority__high high
milestone__ 
branch__branches_doc_trial_output_1384 branches/doc-trial-output-1384
branch_author__Screwtape Screwtape
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__documentation documentation
time__1135949940000000 1135949940000000
changetime__1277833732000000 1277833732000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__spiv cc__warner cc__jml cc__jknight cc__thijs cc__Screwtape cc__khorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
If a test case causes log.err() to be invoked, and then terminates by raising an
exception, the resulting [ERROR] is double counted. Specifically, the summary
line that trial emits as the very last line of output will say:

 FAILED (errors=2, successes=2)

when it ought to say

 FAILED (errors=1, successes=2)

Also the counts don't add up to the number of tests predicted to run (emitted at
the start of the test run).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
I'm not sure how Trial should behave in this instance.

Should Trial report only the final error?  Should Trial report only the first
error?  Should Trial report both, but count them once?  

If the latter, what about the case where there is a log.err() call and the test
fails an assertion? Is it a [FAIL] or an [ERROR]?

Personally, I'd like to change Trial so that log.err() calls do not fail tests.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
This problem isn't only with log.err calls but also with any errors raised in
tearDown or cleanup. 

Trial presently reports errors in tearDown in addition to errors (failures,
unexpected successes, todos etc) in the test method.  For what it's worth, this
is how the stdlib unittest behaves.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
As I understand it, the numbers at the end of a trial run count test methods,
nothing else.

If this is not intended to be the case, then I think it is confusing.  If it is
intended to be the case, then this should be a simple matter of fixing the
counting logic.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
The numbers at the end of a trial run do not count test methods. I don't think
this is confusing.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I'd prefer that the numbers at the end of a trial run show how many methods have
passed and failed. This is useful information for the buildbot to display and
track. I'd be content with a buildbot-specific Reporter having access to this
information (once such a beast is written), but I think humans would benefit
from knowing these numbers too.

The number of errors that occurred during the test run (which may be higher than
the number of tests which failed) may be useful too, although I don't think it
is quite as useful as the number of failed tests. Obviously we want to make both
of them zero, but I think the number of tests which failed is more directly
related to the code under test than the number of errors which occurred.

jml> Personally, I'd like to change Trial so that log.err() calls do not fail tests.

I'm not sure I agree with this. log.err() catches many things that are caused by
the test case but which do not occur on the test method's call stack. The whole
idea of demanding a clean reactor is that everything which might possibly have
been caused by the test method should be completed, and log.err() is the only
way for failures in that activity to be detected. The practice of having
log.err() flunk a test forces people to add error-discarding errbacks to their
Deferreds if that's really what they mean, which is good. It discourages library
code from using log.err() to report non-fatal situations, which has both plusses
and minuses.

Could you elaborate further on how the tests would work without the log.err() check?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
(Note: This discussion is about the summary line, not about the actual reporting
of error tracebacks.)

The number of failures is more directly related to the number of problems to be
fixed.  I really don't see the problem with falling in with the rest of the
xUnit world on this one.  

Buildbot *can* have a custom reporter which behaves difficulty, but it will have
to have an essentially arbitrary algorithm to determine the heading under which
a test gets counted.  (The test FAILs, the tearDown ERRORs -- how do we count
it?)  I recommend that buildbot doesn't do this.

Perhaps the summary line can be supplemented with other information that counts
tests: tests run, tests good (succeeding and skipped) and tests bad (all the rest).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Just as a way of keeping score, I don't think that skipped tests should count as
'good' or 'bad'; since they're not run, we don't know.  I don't want to see
10,000 "good" tests if they're all set to skip :).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
The number of failures has little or nothing to do with the number of problems 
to be fixed. One problem can of course cause many tests to fail, and multiple 
problems could be masked in a single test failure. Having a count of tests that 
fail seems more useful than a count of the number of pieces of tests that fail.

I don't see why it's bad or arbitrary to have
  test fail, tearDown error => ERROR
Currently
  test fails, test errors => ERROR
which makes perfect sense to me.

In any case, however, *I'm* more concerned with the fact that .todo doesn't work 
when the test uses log.err, as per #1397.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
Just to clarify, attached is a test with current output and the suggested outputs.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

I am solidly convinced that this behaviour is a feature and not a bug.  Please re-open the issue if you feel there are new points to be made.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @jesstess<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|Screwtape commented|
|-|-|

It's not exactly clear to me what needs to be done to close this bug. Would it suffice to add a section to the `trial(1)` man page saying something like this?
```
After running the given test suite, the default test reporter prints a summary
of the test run. This consists of the word "PASSED" (if all tests ran as
expected) or "FAILED" (if any test behaved unexpectedly) followed by a count of
the different kinds of test results encountered. The possible kinds of test
results includes:

    successes
	Tests that passed all their assertions and completed without error.
	These are marked "PASSED" in the normal test output.
    failures
	Tests that failed an assertion, called self.fail() or explicitly raised
	self.failureException for some reason. These are marked "FAILED" in the
	normal test output.
    errors
	Tests that raised an unexpected exception (including AssertionError),
	tests that caused the tearDown() method to raise an exception, tests
	that run for longer than the timeout interval, tests that caused
	something to call twisted.python.log.err() without subsequently calling
	self.flushLoggedErrors(), tests that leave the reactor in an unclean
	state, etc. These are marked "ERROR" in the normal test output.
	
	Note that because errors can be caused after the actual test method
	returns, it is possible for a single test to be reported as both an
	error and a failure, and hence the total number of test results can be
	greater than the total number of tests executed.
    skips
	Tests that were skipped, usually because of missing dependencies. These
	are marked "SKIPPED" in the normal test output.
    expectedFailures
	Tests that failed, but were expected to fail, usually because the test
	is for a feature that hasn't been implemented yet. These are marked
	"TODO" in the normal test output.
    unexpectedSuccesses
	Tests that should have been listed under expectedFailures, except that
	for some reason the test succeeded. These are marked "SUCCESS!?!" in
	the normal test output.
```
(assuming such formatting is possible within the constraints of groff and lore)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_18"></a>|@khorn commented|
|-|-|

Wow, whether it would suffice or not, I'd sure like to see something like that someplace in the man page or even in the trial docs.  Nice!

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|Screwtape removed owner|
|-|-|

Buoyed by khorn's enthusiasm, I have attached a patch that adds the sample text of my previous comment to the `trial(1)` man page. Possibly long-form documentation like that should go with the HTML docs, but it seemed easier to work into the text here - also, the HTML docs seem to mostly cover 'how to write tests for trial', and this is more 'how to run trial interactively'.

Notes:
 * It turns out `groff_man(5)` documents all kinds of useful macros for writing consistently formatted man pages that I can't use, because Lore doesn't understand them (sigh).
 * I added a link to the HTML docs in a new "SEE ALSO" section because it seemed appropriate. I'm not sure if it's possible to make the link point to the docs for whatever version of Twisted the user has installed, so I just pointed to the latest version.

I can't find a clearly-defined statement of exactly what conditions are necessary to close this bug, but I believe this patch reasonably covers it, and I assume I have khorn's support for this information being committed somewhere, even if there's more that needs to be written, so -> review.

(hope jesstess doesn't mind me stealing her bug!)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [29442]) Branching to 'doc-trial-output-1384'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [29443]) Apply document-trial-output-1384.patch

refs [#7175](https://github.com/twisted/twisted/issues/7175)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

This looks great to me.  I added a news fragment, too.  I'll merge now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set status to `closed`|
|-|-|

(In [29445]) Merge doc-trial-output-1384

Author: Screwtape
Reviewer: khorn, exarkun
Fixes: [#7175](https://github.com/twisted/twisted/issues/7175)

Document the default trial reporter's summation line in the trial
man page.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
